twitter raw data collection & cron job

nick-c-liu opened the pull request:
- add scripts for twitter raw data collection
- update related readme
- add crontab.bak for the existing cronjob

nick-c-liu committed:
twitter raw data collection & cron job

- add scripts for twitter raw data collection
- update related readme
- add crontab.bak for the existing cronjob

nick-c-liu committed:
Update iffy_get_data.sh

mr-devs commented on crontab.bak:
```diff
@@ -0,0 +1,30 @@
+#####################################################################
+# Twitter Data Collection
+# grab past month twitter data with new iffy list
+0 8 2 * * /home/data/apps/topfibers/repo/scripts/data_collection/iffy_get_data.sh
```
Is there a reason why we are waiting until 8am to do this? Also, is there any reason why we are doing this on the 2nd day of the month instead of the first? I would run this job as soon as we possibly can.

nick-c-liu replied:
Usually, data loading for day `n` would finish during the afternoon of day `n+1`. Since we also wanted to keep our data in compliance and that most deletions happen within 24 hours, it would be nice to wait until the deletion process completes processing deletion data of day `n+1` in the morning of day `n+2`.

If we plug in the last day of the previous month for `n`, we would get 2<sup>nd</sup> for `n+2`.

mr-devs replied:
It seems like this will always be a problem though, right? Any time after we make this call, new deletions could come in that we will have missed for this analysis. Since we are presenting aggregate data and do not show content but simply link to the posts (which will be down as they've been deleted on Twitter), I would prefer to run this as soon as possible.

Am I missing something here?

nick-c-liu replied:
Yeah, I agree. 

Then do you want to run this as soon as the data loading for the last day of the previous month finishes(around 3~4PM) or just 00:01 1<sup>st</sup> day of the month?

I could make the query to get data from the last day of month n-1 to the last day of month n if you want data the first second of the new month, so that we don't miss any data

mr-devs replied:
Yeah, let's do it once we know for sure all of the data we are trying to pull has been loaded to moe. Feel free to buffer in an hour or two to make sure that we know everything is loaded (as I'm assuming that this will take a variable amount of time).

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Please add a docstring explaining what the script is doing more clearly. See [this](https://github.com/mr-devs/top-FIBers/blob/bb51cd05f61c779c150d1a6e5a339a48c125bb04/scripts/data_processing/calc_crowdtangle_fib_all_months.sh) for an example of what I mean.

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,27 @@
+#echo "Tavern Dir: " ${tavern_job}  
+#echo "Moe Key: "${KEY}
+#echo "rsync -at ${fiber_home}iffy_list truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}/"
+#echo "rsync -rt truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}${year_month}/ ${fiber_home}moe_twitter_data/"
```
Please remove these comments as it appears to be old code.

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,27 @@
+#echo "rsync -rt truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}${year_month}/ ${fiber_home}moe_twitter_data/"
+
+# get latest iffy list
+${fiber_home}repo/environments/env_code/bin/python /u/truthy/cronjobs/update_latestmonth_topfibers/iffy_update.py "${fiber_home}iffy_list"
```
Please update this call to use the `iffy_update.py` script in this repository. Right now it is calling w/e is in the `/u/truthy/cronjobs/...` location.

mr-devs replied:
Also, please make sure that the output file is saved [here](https://github.com/mr-devs/top-FIBers/tree/main/data/iffy_files) and the filename is formatted like the file that currently exists in that repo. It should be the date the file is updated like `YYYY-MM-DD`. Currently it is saved in the repo root with the name `iffy_list`.

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,27 @@
+${fiber_home}repo/environments/env_code/bin/python /u/truthy/cronjobs/update_latestmonth_topfibers/iffy_update.py "${fiber_home}iffy_list"
+
+# copy to Lisa-Moe shared drive
+rsync -at ${fiber_home}iffy_list truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}/
```
The above change in the file name will affect this line.

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,27 @@
+rsync -at ${fiber_home}iffy_list truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}/
+
+# create a tavern job
+ssh -i ${KEY} appuser@moe-ln01.soic.indiana.edu "/home/appuser/truthy-cmd.sh get-tweets-with-meme -memes "/mnt/${tavern_job}iffy_list" -tstart "${year_month}-01" -tend "${end_of_last_month}" -tid false -ntweets 1000000 -outdir /mnt/${tavern_job}${year_month}/ -torf false"
```
Same here, the file name will change.

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+year_month=$(date --date='last month' '+%Y-%m')
+end_of_last_month=$(date -d "$(date +%Y-%m-01) -1 day" +%Y-%m-%d)
+
```
Another general note is that we will want to add a logging aspect to this at some point. I don't think it is necessary right now but I will make an issue for the future.

Commit:
add docstring, use argparse

Commit:
use logger to track progress

mr-devs commented on scripts/data_collection/iffy_update.py:
```diff
@@ -0,0 +1,108 @@
+"""
```
Please review this and make sure that I didn't break anything. I tested it a couple of times to see if it works (seems like it does) and if breaking it logs nice notes (seems like it does).

mr-devs replied:
Also, please check that what I've written for the input/output is accurate. I am realizing now that the script may not take the path for the existing file. If it is just the output path for where we will save the new file, please make sure to update the docstring, the `argparse` text, and anything else.

mr-devs reviewed (changes requested):
Thanks for this!! The flow looks good here but I think a few things can be cleaned up (docstrings, old comments/paths). Also, please see the edits that I made to the python script (to ensure that scripts are consistent) and double-check that I didn't break anything.

Please let me know if you have any questions.

nick-c-liu committed:
Update iffy_get_data.sh

update the script to work at the repo root, changed the naming format for new iffy list files downloaded, and added docstring.

nick-c-liu committed:
Update iffy_update.py

adjusted for relative path from repo root, updated comments and docstring

nick-c-liu committed:
Update crontab.bak

update to work at repo root and added log

nick-c-liu reviewed (commented):
Several changes are made to comments, docstring, and code. Most are about using relative path and log

mr-devs commented on scripts/data_collection/iffy_get_data.sh:
```diff
@@ -0,0 +1,50 @@
+rsync -at data/iffy_files/${today}iffy_list.txt truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}${today}iffy_list.txt
+
+# create a tavern job
+ssh -i ${KEY} appuser@moe-ln01.soic.indiana.edu "/home/appuser/truthy-cmd.sh get-tweets-with-meme -memes "/mnt/${tavern_job}${today}iffy_list.txt" -tstart "${year_month}-01" -tend "${end_of_last_month}" -tid false -ntweets 1000000 -outdir /mnt/${tavern_job}${year_month}/ -torf false"
```
Does it matter that the iffy_list files you're passing in this command
```
"/mnt/${tavern_job}${today}iffy_list.txt"
```
is not the one that you just saved in line 44 above 

```
truthy@lisa.luddy.indiana.edu:/home/data/${tavern_job}${today}iffy_list.txt
```
?

Or is it the same somehow?

nick-c-liu replied:
Thanks for going over the changes I made! These are actually pointing toward the same file located on the shared disk between Lisa and Moe. 

prefix "/mnt/" is on Moe and "/home/data/" is on Lisa

mr-devs reviewed (commented):
Everything looks great! Please see my inline comment for one question.

nick-c-liu committed:
updated docstring, added iffy_filename

nick-c-liu reviewed (commented):
I've updated the file to (hopefully) better describe the workflow.

mr-devs commented:
Looks great. thanks!

mr-devs commented:
Please delete the branch, if you don't think you need it anymore.
